### PR TITLE
fix(preset-mini): split shadow rules

### LINKED ISSUE #297: Shadow color not affect
### Shadow color

```html
<div h-full flex justify-center items-center gap-2>
  <div p-14 shadow-lg shadow-red></div> <!-- working -->
  <div p-14 shadow-xl shadow-red></div> <!-- not working -->
</div>
```

Maybe a problem with the order ? 

[demo](https://unocss.antfu.me/?html=DwEwlgbgBAFgtAMwK4BsVQSgpgDygKyQGcAXMBATzgGMsA7ErAJyjEYFsib7GWBzAIYAHOACYAfACgoUUJCgiAjABYoRGAJAB7AO5wUfNRu16mWEOOAB6cBCky50JavWbdcHOlcm4Zi9dspAMhxIA&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmcAvnOlBCHAOQCuWEAxgM6u0BQnqAHpLBQYAhvQA28NJhz5CwIgAoEnOHCjjUrAFxwA2itV7azeqxjUAtOrGpaAGkRxmEMdB20oqZLTIBdA-6kAJRAA&options=N4IgzgLgTglgxhEAuaBXApgXyA)

### PR DESCRIPTION
**Pull request by userquin**
I didn't find any rule with layer, so I'm not sure if this the correct way to fix the issue: what happens if the user provides its own layers with no `color` layer?

Maybe we can change the approach using another internal variable, then initilize it with the default value, for example:

<details>
<summary><strong>packages/preset-mini/src/theme/misc.ts</strong></summary>

```ts
export const boxShadow = {
  'DEFAULT': '0 1px 3px 0 rgba(var(--un-internal-shadow-color), 0.1), 0 1px 2px 0 rgba(var(--un-internal-shadow-color), 0.06)',
  'sm': '0 1px 2px 0 rgba(var(--un-internal-shadow-color), 0.05)',
  'md': '0 4px 6px -1px rgba(var(--un-internal-shadow-color), 0.1), 0 2px 4px -1px rgba(var(--un-internal-shadow-color), 0.06)',
  'lg': '0 10px 15px -3px rgba(var(--un-internal-shadow-color), 0.1), 0 4px 6px -2px rgba(var(--un-internal-shadow-color), 0.05)',
  'xl': '0 20px 25px -5px rgba(var(--un-internal-shadow-color), 0.1), 0 10px 10px -5px rgba(var(--un-internal-shadow-color), 0.04)',
  '2xl': '25px 50px -12px rgba(var(--un-internal-shadow-color), 0.25)',
  'inner': 'inset 0 2px 4px 0 rgba(var(--un-internal-shadow-color), 0.06)',
  'none': 'none',
}
```
</details>

<details>
<summary><strong>packages/preset-mini/src/rules/shadow.ts</strong></summary>

```ts
export const boxShadows: Rule<Theme>[] = [
  [/^shadow-?(.*)$/, ([, d], { theme }) => {
    const value = theme.boxShadow?.[d || 'DEFAULT']
    if (value) {
      return {
        '--un-internal-shadow-color': 'var(--un-shadow-color, 0,0,0)',
        '--un-shadow': value,
        'box-shadow': 'var(--un-ring-offset-shadow, 0 0 #0000), var(--un-ring-shadow, 0 0 #0000), var(--un-shadow)',
      }
    }
    const color = colorResolver(d, theme)
    if (color)
      return color
  }],
]
```
</details>

closes #297

### CODE REVIEW

**Comment by chu121su12** [comment]
Without adding layer, splitting the rules should also prioritize the color.
```js
[/^shadow-(.+)$/, ([, d], { theme }) => colorResolver(d, theme)],
```

Using internal variable is good too. I was thinking using it on border-<direction> to support multiple side coloring.

**Comment by userquin** [comment]
> Without adding layer, splitting the rules should also prioritize the color.
> 
> ```js
> [/^shadow-(.+)$/, ([, d], { theme }) => colorResolver(d, theme)],
> ```
> 
> Using internal variable is good too. I was thinking using it on `border-<direction>` to support individual side coloring.

See the issue, problem with 2 elements.

**Comment by chu121su12** [comment]
When the rules split, the bottom rule will take precedence even without layering.

Snapshot without layer option:
image

**Comment by antfu** [comment]
I don't think we should introduce layers for general rules. Separate the breaking points and color resolvers into two rules with the order will solve the ordering issue.

**Comment by userquin** [comment]
Ok, should we review all rules using default css custom property like this one?

File: packages/preset-mini/src/rules/shadow.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@ const colorResolver = (body: string, theme: Theme) => {
   }
 }
 
+// don't merge rules:
```
Comment (antfu) [inline packages/preset-mini/src/rules/shadow.ts:29]: haha no need for this :)

Comment (antfu) [inline packages/preset-mini/src/rules/shadow.ts:29]:
```suggestion
```

**Pull request merged**